Allow tizen secure storage overwrite

DamianMehers opened the pull request:
### Description of Change ###

Prior to saving a value first try to remove it on tizen so that overwriting an existing value works.

Also fix GetAsync to return `null` if the key does not exist.

### Bugs Fixed ###

https://github.com/xamarin/Essentials/issues/1055
### API Changes ###

None

### Behavioral Changes ###

None

### PR Checklist ###

- [X] Has tests (if omitted, state reason in description)
- [X] Has samples (if omitted, state reason in description): No API surface change
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [X] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc)): No API surface change

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1061) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1061) before we can accept your contribution.<br/><br/>:x: DamianMehers [sign now](https://cla.opensource.microsoft.com/xamarin/Essentials?pullRequest=1061)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/xamarin/Essentials?pullRequest=1061) it.</sub>

DamianMehers commented:
I'm creating this merge requests to allow CI build/test to run since I've not set up my dev machine for Essentials yet.  Underyling bug is still in triage state.

jamesmontemagno commented on Xamarin.Essentials/SecureStorage/SecureStorage.tizen.cs:
```diff
@@ -23,6 +23,14 @@ static Task PlatformSetAsync(string key, string data)
+                try
+                {
+                    // Swallow the exception since we are just removing it if it exists
+                    DataManager.RemoveAlias(key);
```
Is it good to check if it exists first? I would also log Debug message on the catch.

DamianMehers replied:
Re checking first, I wan't sure if we wanted to triple the cost of this call, instead of doubling as it stands - getting it throws an exception if it doesn't exist so you wouild be pretty much back to the current situation.

My thinking is that logging a failed delete will polute the log because it will be happening  all the time as values are overwritten. If I were to call PlatformRemove instead then that would do the log.  This is what I did in the earlier version, but decided to make it silent since it is not actually an error.  If you'd prefer I can log a message though - I'm not sure it is actionable.

DamianMehers replied:
I'm open to implementing either of your suggestions after you've considered my reasoning.  Let me know.

Redth replied:
Exceptions aren't necessarily cheap either, so my gut says we should probably check before removing.

DamianMehers replied:
> Exceptions aren't necessarily cheap either, so my gut says we should probably check before removing.

I know - that is why I'm not sure that checking makes sense because checking throws an exception if the key is not there.  I'll update the code as you suggest though :-)

DamianMehers replied:
Let me know if you'd prefer that I use the native calls rather than the platform wrappers.

DamianMehers replied:
This won't work, since the Tizen `PlatformGetAsync` looks to throw an Exception for non-existent values (a separate bug).  I'll update the code again.

jamesmontemagno commented:
You can grab and test the nuget from the CI build: https://dev.azure.com/xamarin/public/_build/results?buildId=12699&view=artifacts&type=publishedArtifacts

DamianMehers commented:
@jamesmontemagno @Redth it seems to me that the Tizen `PlatformGetAsync` is broken too because it throws an exception if the key does not exist, rather than returning null.

Would you prefer that I create a new bug/PR, or update this PR to add a new test which does a 

``` c#
PlatformGetAsync(Guid.New.Guid().ToString());
```
and fix the Tizen `PlatformGetAsync` to return `null` if the underlying `DataManager.Get` throws `InvalidOperationException`?

Redth commented:
Hmmm is there any method on DataManager to check if the key exists?

If not yeah I guess we need to catch the Get.  Same PR is fine if you’re willing to do the work, thanks!

DamianMehers commented:
@Redth @jamesmontemagno ... ready for a new review. 

I've fixed the `GetAsync` to return null when the key does not exist, and  I've added comments to the code to clarify Tizen's behavior - the Tizen `DataManager` API has no way to check for a key's existence.

DamianMehers commented:
> Thanks for figuring out the behaviour here and submitting the PR!

Thanks for approving.  I'll pull in the nuget to the app I'm building (A Trello client for Tizen - I've already got it running on WatchOS and Wear OS thanks to Xamarin) and double-check.

ppaneksamsung commented:
@DamianMehers I tried to use SecureStorage API on Tizen. I'm not sure if your intention was to return a null Task, when a key does not exist. In other words:

`var value = await SecureStorage.GetAsync("SomeNotExistingKey");`

throws NullReferenceException, while I would expect that a task completes with no exception and a null value is returned. Is it intentional? If not, then probably:

```
            catch (System.InvalidOperationException)
            {
                // The DataManager.Get call throws an exception if key does not exist. Not logging
                // anything since this is an expected and normal situation (if the key did not exist).
                return null;
            }
```

needs to be changed to something like:

```
            catch (System.InvalidOperationException)
            {
                // The DataManager.Get call throws an exception if key does not exist. Not logging
                // anything since this is an expected and normal situation (if the key did not exist).
                return Task.FromResult<string>(null);
            }
```

mattleibow commented:
@ppaneksamsung is correct, let me fix that. Thanks for spotting it!

DamianMehers commented:
> @DamianMehers I tried to use SecureStorage API on Tizen. I'm not sure if your intention was to return a null Task, when a key does not exist. In other words:
> 
> `var value = await SecureStorage.GetAsync("SomeNotExistingKey");`
> 
> throws NullReferenceException, while I would expect that a task completes with no exception and a null value is returned. Is it intentional? If not, then probably:
> 
> ```
>             catch (System.InvalidOperationException)
>             {
>                 // The DataManager.Get call throws an exception if key does not exist. Not logging
>                 // anything since this is an expected and normal situation (if the key did not exist).
>                 return null;
>             }
> ```
> 
> needs to be changed to something like:
> 
> ```
>             catch (System.InvalidOperationException)
>             {
>                 // The DataManager.Get call throws an exception if key does not exist. Not logging
>                 // anything since this is an expected and normal situation (if the key did not exist).
>                 return Task.FromResult<string>(null);
>             }
> ```

Yep - dumb bug on my part.  Apologies.
